demo/python-subtitle/main.py

```diff
@@ -0,0 +1,156 @@
+#
+#    Copyright 2022 Picovoice Inc.
+#
+#    You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
+#    file accompanying this source.
+#
+#    Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+#    an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+#    specific language governing permissions and limitations under the License.
+#
+
+import os
+import sys
+import time
+from argparse import ArgumentParser
+from threading import Thread
+from typing import *
+
+import pvleopard
+from pytube import YouTube
+
+
+class ProgressAnimation(Thread):
+    def __init__(self, prefix: str, step_sec: float = 0.19) -> None:
+        super().__init__()
+
+        self._prefix = prefix
+        self._step_sec = step_sec
+        self._frames = (
+            ".  ",
+            ".. ",
+            "...",
+            " ..",
+            "  .",
+            "   "
+        )
+        self._stop = False
+
+    def run(self) -> None:
+        while True:
+            for frame in self._frames:
+                if self._stop:
+                    sys.stdout.write('\r%s\r' % (" " * (len(self._prefix) + 1 + len(frame))))
+                    self._stop = False
+                    return
+                sys.stdout.write('\r%s %s' % (self._prefix, frame))
+                time.sleep(self._step_sec)
+
+    def stop(self) -> None:
+        self._stop = True
+        while self._stop:
+            pass
+
+
+def second_to_timecode(x: float) -> str:
+    hour = int(x) // 3600
+    x -= hour * 3600
+    minute = int(x) // 60
+    x -= minute * 60
+    second = int(x)
+    x -= second
+    millisecond = int(x * 1000.)
+
+    return '%.2d:%.2d:%.2d,%.3d' % (hour, minute, second, millisecond)
+
+
+def to_srt(words: Sequence[pvleopard.Leopard.Word], endpoint_sec: float = 1., length_limit: Optional[int] = 16) -> str:
+    def _helper(start: int, end: int) -> None:
+        lines.append("%d" % section)
+        lines.append("%s --> %s" % (second_to_timecode(words[start].start_sec), second_to_timecode(words[end].end_sec)))
+        lines.append(' '.join(x.word for x in words[start:(end + 1)]))
+        lines.append('')
+
+    lines = list()
+    section = 0
+    j = -1
+    for k in range(1, len(words), 1):
+        if ((words[k].start_sec - words[k - 1].end_sec) >= endpoint_sec) or \
+                (length_limit is not None and (k - j - 1) >= length_limit):
+            _helper(start=(j + 1), end=(k - 1))
+            j = k - 1
```

Comment:
We set `j` by subtracting 1 from `k`, but then we ever only use `j + 1`. It would be cleaner to shift `j` by one, i.e., initialize it as 0 and remove the 1s in the other lines. Then we could even rename it to `start` and no longer pass it to `_helper` as an argument, which could help with readability.

Reply:
agreed